### PR TITLE
add override implementation for testing framework

### PR DESCRIPTION
**Pull request by ollevche**
Resolves https://github.com/opentofu/opentofu/issues/1204

This PR adds initial implementation for `override_resource`, `override_data` and `override_module` blocks for testing framework.

## Implementation details

Overriding data and managed resources is implemented by avoiding calls to the underlying provider methods: `ReadResource`, `PlanResourceChange`, `ApplyResourceChange`, and `ReadDataSource`. It's implemented via a wrapper of the Provider interface, which is injected in a constructor (node's `getProvider` function). The wrapper has additional knowledge of when / why it's called so it mocks needed resources / data in a format of one-time use. It will be extended later to cover `mock_provider` and `mock_resource` / `mock_data` implementation.

Overriding modules is implemented by not adding a module configuration to the DAG and making the output nodes use static values from `override_module` block instead of evaluating the original expression.

## Reasoning on testing

The implementation requires a lot of small changes across the codebase and the only large chunk of code is actual mock generation. I created an e2e test to ensure the override feature works well in general. It could be extended later to test similar behavior of `mock_providers` and `mock_resource`. Unit tests cover the mock generation mechanism to ensure attributes and blocks are generated correctly for different edge cases. Also, newly created addr parsing functions are also covered with unit tests and can be reused later on.

## TransformForTest note

I think the `TransformForTest` function is a bit hard to extend so we would need to consider minor refactoring in the case of a major extension in the future.

## Target Release

1.8.0

### CODE REVIEW

File: internal/addrs/module.go [hunk 1]
```diff
@@ -177,6 +177,25 @@ func (m Module) configRemovableSigil() {
...
 }
 
+// ParseModule parses a module address from the given traversal,
+// which has to contain only the module address with no resource/data/variable/etc.
+// This function only supports module addresses without instance keys (as the
+// returned Module struct doesn't support instance keys) and will return an
+// error if it encounters one.
+func ParseModule(traversal hcl.Traversal) (Module, tfdiags.Diagnostics) {
+	mod, remain, diags := parseModulePrefix(traversal)
+	if len(remain) != 0 {
```
Comment (cam72cam) [inline internal/addrs/module.go:187]: Perhaps also check that `!diags.HasErrors()` ?

When diags.HasErrors(), any results from a function should be considered undefined.

Comment (cam72cam) [reply internal/addrs/module.go:187]: I see you did that for `ParseConfigResource` below

Comment (ollevche) [reply internal/addrs/module.go]: Yep, missed this one, thank you! I was still getting used to diags instead of errors by the time I was writing this line :)

Fixed in ff1754d396ecae51c1ac70046647dc7b880e2e13

File: internal/configs/test_file.go [hunk 2]
```diff
@@ -537,6 +654,181 @@ func decodeTestRunOptionsBlock(block *hcl.Block) (*TestRunOptions, hcl.Diagnosti
 	return &opts, diags
 }
 
+func decodeOverrideResourceBlock(block *hcl.Block, mode addrs.ResourceMode) (*OverrideResource, hcl.Diagnostics) {
```
Comment (cam72cam) [inline internal/configs/test_file.go:657]: It looks like there's some duplication between the next few functions that could be refactored out?

Comment (ollevche) [reply internal/configs/test_file.go:657]: Yep, there's an almost identical function that could be moved out. Thanks!

[8b7f372](https://github.com/opentofu/opentofu/pull/1499/commits/8b7f372415f9a4dccbfad433212f9a6a5fbaa1bf)

File: internal/configs/config.go [hunk 3]
```diff
@@ -955,8 +985,193 @@ func (c *Config) TransformForTest(run *TestRun, file *TestFile) (func(), hcl.Dia
...
+		resources = append(resources, r)
+	}
+
+	return resources, diags
+}
+
+func mergeOverridenModules(runModules, fileModules []*OverrideModule) ([]*OverrideModule, hcl.Diagnostics) {
+	// runAddrs is a unique set of module addresses in run block.
+	// It's already validated for duplicates previously.
+	runAddrs := make(map[string]struct{})
```
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/config.go:1150]: It stands for module addresses in run block. It could be `runModuleAddrs` or `addrsInRun` if just `runAddrs` is not clear

File: internal/configs/config.go [hunk 4]
```diff
@@ -955,8 +985,193 @@ func (c *Config) TransformForTest(run *TestRun, file *TestFile) (func(), hcl.Dia
...
+				output.OverrideValue = nil
+			}
+		}
+	}, diags
+}
+
+func mergeOverridenResources(runResources, fileResources []*OverrideResource) ([]*OverrideResource, hcl.Diagnostics) {
+	// runAddrs is a unique set of resource addresses in run block.
+	// It's already validated for duplicates previously.
+	runAddrs := make(map[string]struct{})
```
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/config.go:1118]: The same as in `mergeOverridenModules`: It stands for resource addresses in run block. It could be `runResourceAddrs` or `addrsInRun` if just `runAddrs` is not clear

File: internal/configs/resource.go [hunk 5]
```diff
@@ -49,6 +50,14 @@ type Resource struct {
 	// If this is nil, then this resource is essentially public.
 	Container Container
 
+	// IsOverridden indicates if the resource is being overriden. It's used in
+	// testing framework to not call the underlying provider.
+	IsOverridden bool
```
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/resource.go:55]: Yep, it could be done this way. We will need to check for a nil map value (not overridden), a zero-length map (overridden with no values), and a filled map (overridden with values). I just find it harder to understand that way from the reader's perspective.

Also, I think it's easy to misuse nil and zero-length maps this way so an additional flag comes in handy.

[the remark this answers is not in the document]
Comment (siddharthasonker95) [reply internal/configs/resource.go:55]: that makes sense... just wanted to know the idea behind this.
Great work Thanks!

File: internal/configs/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/config.go]: 7e162882a7c5616aa71cd116cc67197bd5d997d7

File: internal/configs/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ollevche) [reply internal/configs/config.go]: 7e162882a7c5616aa71cd116cc67197bd5d997d7

**Comment by ollevche** [comment]
I purposely left some failing lint checks to discuss if we need to tweak its settings. These functions looks fine to me so I suggest to increase the limit of complexity check. @opentofu/core-engineers what do you think?

File: internal/configs/test_file.go [hunk 6]
```diff
@@ -537,6 +662,154 @@ func decodeTestRunOptionsBlock(block *hcl.Block) (*TestRunOptions, hcl.Diagnosti
...
+	if attr, exists := content.Attributes["outputs"]; exists {
+		outputs, moreDiags := parseObjectAttrWithNoVariables(attr)
+		mod.Outputs, diags = outputs, append(diags, moreDiags...)
+	}
+
+	return mod, diags
+}
+
+func parseObjectAttrWithNoVariables(attr *hcl.Attribute) (map[string]cty.Value, hcl.Diagnostics) {
+	if vars := attr.Expr.Variables(); len(vars) != 0 {
```
Comment (cam72cam) [inline internal/configs/test_file.go:738]: Is this explicitly needed? The Expr.Value(nil) below should produce a similar diagnostic?

**Comment by cam72cam** [comment]
Current Thoughts:
1) Is there address parsing code that could be re-used from import/moved/removed blocks?
2) How does this handle for_each/count interactions?  Can we get a test for this?
3) All of the new addr functions should have comprehensive unit tests.
4) The mockValueComposer needs to be moved to it's own file, possibly in the hcl2shim package where there's already similar functionality?

The module node mocking seems reasonable, but the resource node mocking feels tacked on due to all of the special casing.  I know we talked about this in the past and I think we should think about potential alternative solutions here.  We might be able to inject a custom provider that only knows how to handle that specific resource instance?  This could also be useful when we get to provider mocking.

Overall I think that this work makes sense and is really good so far!

File: internal/configs/test_file.go (hunk 6 above)
Comment (ollevche) [reply internal/configs/test_file.go]: Agree, removed this call in c398e309df9d282bf765ba5578fef6573de032c0

**Comment by ollevche** [comment]
@cam72cam thank you for your thoughts and suggestions!

1. I can't reuse public functions directly, since they are specific to the concrete block (i.e. `ParseMoveEndpoint`, `ParseRemoveEndpoint`) and also I can't reuse `ParseAbsResourceInstance` (which is used in import) because it allows instance addresses. So I added public counterparts for existing private functions (`addrs.ParseConfigResource`, `addrs.ParseModule`) to parse resource and module addresses and return existing public structs concerning their internal logic (like disallow instance address). I think this way it could be better reused in the future. WDYT?

2. We mock each existing instance of resource / data / module. If there's also values / outputs field present to use in mocking - we will use it for each instance. I'll extend the e2e test to include such a scenario. Agree.

3. In the case, that we are moving with the current implementation, I'll add unit tests for these functions. Agree.

4. Yep, the mocking code needs to be moved somewhere else. Agree.

5. Yep, we discussed the option of custom provider implementation earlier. I will reconsider this approach once more!

**Review by cam72cam — APPROVED** [review]

**Review by cam72cam — DISMISSED** [review]

**Review by cam72cam — DISMISSED** [review]

File: internal/configs/hcl2shim/mock_value_composer.go [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+			))
+		}
+	}
+
+	mockAttrs := make(map[string]cty.Value)
+
+	impliedTypes := schema.ImpliedType().AttributeTypes()
+
+	for k, attr := range schema.Attributes {
+		// If the value present in configuration - just use it.
```
Comment (Evi1Pumpkin) [inline internal/configs/hcl2shim/mock_value_composer.go:94]: I want to validate I understand this scenario correctly -
From what I understand if I have the resource:
```
resource "something" "hi" {
    id = "2342"
}
```
If I have an "override_resource" block for this resource, I won't be able to override the `id` attribute?

File: internal/configs/hcl2shim/mock_value_composer.go [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+			mockAttrs[k] = cv
+			addPotentialDefaultsWarning(k, "The field is ignored since overriding configuration values not allowed.")
+			continue
+		}
+
+		// Non-computed attributes can't be generated
+		// so we set them from configuration only.
+		if !attr.Computed {
+			mockAttrs[k] = cty.NullVal(attr.Type)
+			addPotentialDefaultsWarning(k, "The field is ignored since overriding non-computed fields not allowed.")
```
Comment (Evi1Pumpkin) [inline internal/configs/hcl2shim/mock_value_composer.go:105]:
```suggestion
			addPotentialDefaultsWarning(k, "The field is ignored since overriding non-computed fields is not allowed.")
```

File: internal/configs/hcl2shim/mock_value_composer.go [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+	mockAttrs := make(map[string]cty.Value)
+
+	impliedTypes := schema.ImpliedType().AttributeTypes()
+
+	for k, attr := range schema.Attributes {
+		// If the value present in configuration - just use it.
+		if cv, ok := configMap[k]; ok && !cv.IsNull() {
+			mockAttrs[k] = cv
+			addPotentialDefaultsWarning(k, "The field is ignored since overriding configuration values not allowed.")
```
Comment (Evi1Pumpkin) [inline internal/configs/hcl2shim/mock_value_composer.go:97]:
```suggestion
			addPotentialDefaultsWarning(k, "The field is ignored since overriding configuration values is not allowed.")
```

File: internal/configs/hcl2shim/mock_value_composer.go [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+			}
+		}
+
+		// If there's no value in defaults, we generate our own.
+		v, ok := mvc.getMockValueByType(impliedTypes[k])
+		if !ok {
+			diags = diags.Append(tfdiags.WholeContainingBody(
+				tfdiags.Error,
+				"Failed to generate mock value",
+				fmt.Sprintf("Mock value cannot be generated for dynamic type. Please, specify `%v` field explicitly in configuration.", k),
```
Comment (Evi1Pumpkin) [inline internal/configs/hcl2shim/mock_value_composer.go:133]:
```suggestion
				fmt.Sprintf("Mock value cannot be generated for dynamic type. Please specify the `%v` field explicitly in the configuration.", k),
```

File: internal/configs/test_file.go [hunk 11]
```diff
@@ -537,6 +662,143 @@ func decodeTestRunOptionsBlock(block *hcl.Block) (*TestRunOptions, hcl.Diagnosti
...
+		return traversal, &configRes, diags
+	}
+
+	res := &OverrideResource{
+		Mode: mode,
+	}
+
+	content, diags := block.Body.Content(overrideResourceBlockSchema)
+
+	if attr, exists := content.Attributes["target"]; exists {
```
Comment (Evi1Pumpkin) [inline internal/configs/test_file.go:688]: Isn't the `target` attribute in the block required? Feels like we should throw an error id it's missing/

File: internal/configs/test_file.go [hunk 12]
```diff
@@ -537,6 +662,143 @@ func decodeTestRunOptionsBlock(block *hcl.Block) (*TestRunOptions, hcl.Diagnosti
...
+		}
+
+		return traversal, target, diags
+	}
+
+	mod := &OverrideModule{}
+
+	content, diags := block.Body.Content(overrideModuleBlockSchema)
+
+	if attr, exists := content.Attributes["target"]; exists {
```
Comment (Evi1Pumpkin) [inline internal/configs/test_file.go:723]: Same as before - Isn't the `target` attribute in the block required? Feels like we should throw an error if it's missing.

Comment (ollevche) [reply internal/configs/test_file.go:723]: Same - Yes, it is required and we throw an error if it's missing. The check is done before this function is called.

It is marked as required in the schema: https://github.com/opentofu/opentofu/blob/override-for-testing/internal/configs/test_file.go#L893

File: internal/configs/test_file.go (hunk 11 above)
Comment (Evi1Pumpkin) [reply internal/configs/test_file.go:688]: Awesome, missed that :)

File: internal/tofu/provider_for_test_framework.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func (p providerForTest) PlanResourceChange(r providers.PlanResourceChangeRequest) providers.PlanResourceChangeResponse {
+	if r.Config.IsNull() {
+		return providers.PlanResourceChangeResponse{
+			PlannedState: r.ProposedNewState, // null
+		}
+	}
+
+	if p.plannedChange != nil {
```
Comment (Evi1Pumpkin) [inline internal/tofu/provider_for_test_framework.go:52]: I don't really understand this part of the `planned change` and how it affects the overridden resource. Can you explain a little?

**Review by Evi1Pumpkin — APPROVED** [review]
Splendid job! This is a really useful feature 🥇

File: internal/tofu/provider_for_test_framework.go (hunk 13 above)
Comment (ollevche) [reply internal/tofu/provider_for_test_framework.go:52]: Sure!

In general, `providerForTest` uses configuration values (from `values` / `outputs` field in `override_` block) and automatically generates some missing computed values. When doing so, it uses some rules on how the values should be generated and the value is random for strings.

After this, tofu checks if the provider generated the correct response. It's the point when the logic breaks without this `if`. We need to use the same generated value for multiple sequential provider calls (i.e. plan is called twice).

Currently, `providerForTest` is meant to be used once inside the `NodeAbstractResourceInstance` so we just pass the previously planned value to the `providerForTest`.

It will be changed to the proper cache when extending the `providerForTest` to support `mock_provider` feature. This one: https://github.com/opentofu/opentofu/issues/1155

I left a comment explaining it a bit more here: https://github.com/opentofu/opentofu/blob/override-for-testing/internal/tofu/provider_for_test_framework.go#L21

Comment (Evi1Pumpkin) [reply internal/tofu/provider_for_test_framework.go:52]: Thanks for the explanation, this is much clearer! 🙏
